### PR TITLE
Release script for FinniversKit and FinnUI

### PR DESCRIPTION
**Pull request by osanoj**
# Why?
To make it harder to do a mistake when making a new release.

# What?
After a few failed/complicated attempts with Swift and shell scripts I made a fastlane version that seems a lot better.

There will be a few questions when running the release scripts, mainly to allow you to do a "dry-run", so you will be able to choose if the branch will be pushed and a release created.

It is intended to be used from the master branch and to push the updated podspec directly to master branch. When the script is done it will open the release in the browser so you can edit the description.

I added `fastlane` to the `Gemfile`, don't forget to `bundle install`!

Finally to make a release with this script you need to create a personal access token in GitHub and add it to a "FINNIVERSKIT_GITHUB_ACCCESS_TOKEN" environment variable, like for our main app (since you need to use the API to create a new release). 

# Show me
This is from a `bundle exec fastlane` test run I made (I could also have used `bundle exec fastlane make_finnui_version bump:patch`):
```
[16:41:25]: Welcome to fastlane! Here's what your app is set up to do:
+--------+--------------------------+-------------------------------------------+
|                            Available lanes to run                             |
+--------+--------------------------+-------------------------------------------+
| Number | Lane Name                | Description                               |
+--------+--------------------------+-------------------------------------------+
| 1      | make_finniverkit_version | Create a new FinniversKit version         |
| 2      | make_finnui_version      | Create a new FinnUI version               |
| 3      | verify_ssh_to_github     | Attempt to connect to github.com with ssh |
| 0      | cancel                   | No selection, exit fastlane!              |
+--------+--------------------------+-------------------------------------------+
[16:41:25]: Which number would you like run?
2
[16:41:31]: Running lane `make_finnui_version`. Next time you can do this by directly typing `fastlane make_finnui_version` 🚀.
[16:41:31]: Driving the lane 'make_finnui_version' 🚀
[16:41:31]: ------------------------
[16:41:31]: --- Step: git_branch ---
[16:41:31]: ------------------------
[16:41:31]: Not on 'master' branch, are you sure you know what you are doing? (y/n)
y
[16:41:34]: On 'release-script' not 'master' branch!

[16:41:34]: ------------------------
[16:41:34]: --- Step: git_branch ---
[16:41:34]: ------------------------
[16:41:34]: What do you want to bump: 
1. major
2. minor
3. patch
?  3
[16:41:36]: ---------------------------------
[16:41:36]: --- Step: version_get_podspec ---
[16:41:36]: ---------------------------------
[16:41:36]: ----------------------------
[16:41:36]: --- Step: git_tag_exists ---
[16:41:36]: ----------------------------
[16:41:38]: Shell command exited with exit status 2 instead of 0.
[16:41:38]: Will create a release with verstion '1.2.1' and tag 'finnui-1.2.1'. Is that correct? (y/n)
y
[16:41:40]: ----------------------------------
[16:41:40]: --- Step: version_bump_podspec ---
[16:41:40]: ----------------------------------
[16:41:40]: Will commit changes to files (y/n)
y
[16:41:44]: ------------------------
[16:41:44]: --- Step: git_commit ---
[16:41:44]: ------------------------
[16:41:44]: $ git commit -m Bumped\ FinnUI\ version\ to\ 1.2.1 FinnUI.podspec
[16:41:44]: ▸ [release-script 1b84db60] Bumped FinnUI version to 1.2.1
[16:41:44]: ▸ 1 file changed, 1 insertion(+), 1 deletion(-)
[16:41:44]: Successfully committed "FinnUI.podspec" 💾.
[16:41:44]: Will push branch changes to GitHub (y/n)
y
[16:41:45]: --------------------------------
[16:41:45]: --- Step: push_to_git_remote ---
[16:41:45]: --------------------------------
[16:41:45]: $ pwd
[16:41:45]: ▸ /Users/jonas_work/develop/FinniversKit
[16:41:45]: $ git push origin release-script:release-script --tags
[16:41:50]: ▸ To github.com:finn-no/FinniversKit.git
[16:41:50]: ▸ 5002fb42..1b84db60  release-script -> release-script
[16:41:50]: Successfully pushed to remote.
[16:41:50]: Will create a release in GitHub with tag 'finnui-1.2.1' on branch 'release-script' (y/n)
y
[16:41:52]: --------------------------------
[16:41:52]: --- Step: set_github_release ---
[16:41:52]: --------------------------------
[16:41:52]: Creating release of finn-no/FinniversKit on tag "finnui-1.2.1" with name "FinnUI 1.2.1".
[16:41:54]: Successfully created release at tag "finnui-1.2.1" on GitHub
[16:41:54]: See release at "https://github.com/finn-no/FinniversKit/releases/tag/finnui-1.2.1"
[16:41:54]: Updated version and created release on Github!
[16:41:54]: Please update its description here: https://github.com/finn-no/FinniversKit/releases/tag/finnui-1.2.1
[16:41:54]: ----------------------
[16:41:54]: --- Step: git_pull ---
[16:41:54]: ----------------------
[16:41:54]: $ git fetch --tags
[16:41:55]: ▸ From github.com:finn-no/FinniversKit
[16:41:55]: ▸ * [new tag]           finnui-1.2.1 -> finnui-1.2.1
```

I have already deleted that release, tag and commit btw 😛

### CODE REVIEW

**Comment by osanoj** [comment]
I will be adding some documentation for this, but you can have a look at the solution now

File: fastlane/Fastfile [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+  def make_release(project, tag_prefix, options)
+    ## Verify the branch.
+    branch = verify_branch("master")
+
+    ## Verify clean git status.
+    #ensure_git_status_clean
+
+    ## Create the new tag.
+    # If no bump in version is specified, only the release type version will be bumped. Ie. from 'rc1' to 'rc2'.
```
Comment (bstien) [inline fastlane/Fastfile:40]: Should probably change this line since the `rc1` part is not applicable here

File: fastlane/Fastfile [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    sshresult = result
+    if sshresult.include? "successfully authenticated"
+      UI.success "Could authenticate to #{githuburl}"
+    else
+      UI.shell_error! "Seems connection to #{githuburl} failed"
+    end
+    })
+  end
+
+  def make_release(project, tag_prefix, options)
```
Comment (bstien) [inline fastlane/Fastfile:32]: Just to clear things up, the methods below should be moved out of the lane `verify_ssh_to_github` and into the global scope instead

File: fastlane/Fastfile [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    ## Verify the branch.
+    branch = verify_branch("master")
+
+    ## Verify clean git status.
+    #ensure_git_status_clean
```
Comment (bstien) [inline fastlane/Fastfile:37]: These lines can be removed

File: fastlane/Fastfile [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+    ## Make the user verify that the tag is correct.
+    unless UI.confirm "Will create a release with verstion '#{new_version}' and tag '#{new_tag}'. Is that correct?"
+      UI.user_error! "Version was not accepted by user."
+    end
+
+    version_bump_podspec(path: "#{project}.podspec", version_number: new_version)
+
+    ## Update the version used when building
+    #todo: update_version_file("../#{project}/Version.h", new_version)
```
Comment (bstien) [inline fastlane/Fastfile:59]: Are these ones needed?

File: fastlane/Fastfile [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+      UI.user_error! "Tag '#{tag}' already exists! Check github for the full list of tags. Bump the version if necessary."
+    end
+  end
+
+  def create_github_release(title, tag, branch)
+    return set_github_release(
+      repository_name: "finn-no/FinniversKit",
+      name: title,
+      tag_name: tag,
+      api_token: github_token,
```
Comment (bstien) [inline fastlane/Fastfile:153]: Just noticed this one now. I'm sure we do the same for the app, but what happens if this envvar isn't set before you run the script?

Maybe `verify_env_var` should be called before the `make_release` function within the `make_x_version` lanes? Just so we're sure the envvar is set before we do any work at all. It sucks to do all the steps just to be denied when trying to push the release 😄

File: fastlane/Fastfile (hunk 2 above)
Comment (osanoj) [reply fastlane/Fastfile:32]: Hehe, there was an indent error, fixed it, thanks 👍

File: fastlane/Fastfile (hunk 1 above)
Comment (osanoj) [reply fastlane/Fastfile:40]: I see now I should have read through the comments that were in the code I copied 😅

File: fastlane/Fastfile (hunk 4 above)
Comment (osanoj) [reply fastlane/Fastfile:59]: No, not really, but my plan is to create these files so the build of the framework will actually specify the same version as the podspec, but can remove it until then

File: fastlane/Fastfile (hunk 5 above)
Comment (osanoj) [reply fastlane/Fastfile:153]: Yes! Moving it to the very top 👍

**Pull request merged**